### PR TITLE
ports: allow base port number to be configured

### PR DESCRIPTION
**Pull request by andrewbonney**
It would be useful to be able to run the tool as multiple instances on the same host. To do this I need to offset the port ranges used. Use of `ENABLE_DNS_SD` when running multiple instances won't be possible, but it'll help to test things other than discovery.

### CODE REVIEW

**Comment by garethsb-sony** [comment]
To support this, we need to make service instance names and port numbers in https://github.com/AMWA-TV/nmos-testing/blob/master/test_data/IS0401/dns_records.zone configurable, and port numbers in https://github.com/AMWA-TV/nmos-testing/blob/master/test_data/BCP00301/ca/openssl.cnf.intermediate as well, I think?

**Comment by andrewbonney** [comment]
Technically yes, but both cases are a pain for different reasons. The config file notes that they'd need changing manually. I could always note instead that certain tests won't work if you change the PORT_BASE for now (ie. discovery and BCP-003-01 CRL/OCSP).

**Comment by garethsb-sony** [comment]
Oh... Obviously, I didn't read that :-/

**Comment by andrewbonney** [comment]
DNS port changes will now happen automatically. I'm leaving the OpenSSL config as it is because changes to that will also require re-generation of certificates.

File: test_data/IS0401/dns_records.zone [hunk 1]
```diff
@@ -3,84 +3,84 @@ mocks.{{ domain }}.	IN	A	{{ ip_address }}
 timeout.{{ domain }}.	IN	A	192.0.2.1
 
 ; There should be one PTR record for each instance of the service you wish to advertise.
-_nmos-registration._tcp	PTR	reg-api-5101-ver._nmos-registration._tcp
-_nmos-register._tcp	PTR	reg-api-5101-ver._nmos-register._tcp
-_nmos-query._tcp	PTR	qry-api-5101-ver._nmos-query._tcp
-_nmos-registration._tcp	PTR	reg-api-5101-proto._nmos-registration._tcp
-_nmos-register._tcp	PTR	reg-api-5101-proto._nmos-register._tcp
-_nmos-query._tcp	PTR	qry-api-5101-proto._nmos-query._tcp
+_nmos-registration._tcp	PTR	reg-api-101-ver._nmos-registration._tcp
```
Comment (garethsb-sony) [inline test_data/IS0401/dns_records.zone]: Why won't the service instance names like `reg-api-101...` conflict between multiple test suite instances running on the same host?

**Comment by garethsb-sony** [comment]
> I'm leaving the OpenSSL config as it is because changes to that will also require re-generation of certificates.

Makes sense!

File: test_data/IS0401/dns_records.zone (hunk 1 above)
Comment (andrewbonney) [reply test_data/IS0401/dns_records.zone]: The DNS server can only run on port 53 as I don't believe Nodes could be configured to find it anywhere else. As such you're only ever going to be able to run one unicast DNS instance on a host at a time.

**Comment by garethsb-sony** [comment]
Can’t decide whether this is worthwhile or not… I think the faff with the port logic makes it not, but you decide!

```DNS Zone
; These lines give the fully qualified DNS names to the IP addresses of the hosts which we'd like to discover
mocks.{{ domain }}.	IN	A	{{ ip_address }}
timeout.{{ domain }}.	IN	A	192.0.2.1

{% for inst in ['1-ver', '1-proto', '2', '3', '4', '5', 'timeout', '6'] %}

; There should be one PTR record for each instance of the service you wish to advertise.
_nmos-registration._tcp	PTR	reg-api-{{ inst }}._nmos-registration._tcp
_nmos-register._tcp	PTR	reg-api-{{ inst }}._nmos-register._tcp
_nmos-query._tcp	PTR	qry-api-{{ inst }}._nmos-query._tcp

; Next we have a SRV and a TXT record corresponding to each PTR above, first the Registration API
; The SRV links the PTR name to a resolvable DNS name (see the A records above) and identify the port which the API runs on
; The TXT records indicate additional metadata relevant to the IS-04 spec

{% if inst in ['1-ver', '1-proto'] %}
{% set port = reg_port_base + 1 %}
{% elif inst in ['2', '3', '4', '5'] %}
{% set port = reg_port_base + loop.index0 %}
{% elif inst in ['timeout'] %}
{% set port = 444 %}
{% elif inst in ['6'] %}
{% set port = reg_port_base + 6 %}
{% endif %}
{% set host = 'mocks' if inst != 'timeout' else 'timeout' %}
{% set ver = api_ver if inst != '1-ver' else 'v9.0' %}
{% set proto = api_proto if inst != '1-proto' else 'invalid' %}
{% set pri = loop.index0 * 10 %}

reg-api-{{ inst }}._nmos-registration._tcp	SRV	0 0 {{ port }} {{ host }}.{{ domain }}.
reg-api-{{ inst }}._nmos-register._tcp	SRV	0 0 {{ port }} {{ host }}.{{ domain }}.
reg-api-{{ inst }}._nmos-registration._tcp	TXT	"api_ver={{ ver }}" "api_proto={{ proto }}" "pri={{ pri }}"
reg-api-{{ inst }}._nmos-register._tcp	TXT	"api_ver={{ ver }}" "api_proto={{ proto }}" "pri={{ pri }}"

; Finally, the SRV and TXT for the Query API
qry-api-{{ inst }}._nmos-query._tcp	SRV	0 0 {{ port }} {{ host }}.{{ domain }}.
qry-api-{{ inst }}._nmos-query._tcp	TXT	"api_ver={{ ver }}" "api_proto={{ proto }}" "pri={{ pri }}"

{% endfor %}
```

**Comment by garethsb-sony** [comment]
Or even loopier...

```DNS Zone
; These lines give the fully qualified DNS names to the IP addresses of the hosts which we'd like to discover
mocks.{{ domain }}.	IN	A	{{ ip_address }}
timeout.{{ domain }}.	IN	A	192.0.2.1

{% for inst, idx in [('1-ver', 1), ('1-proto', 1), ('2', 2), ('3', 3), ('4', 4), ('5', 5), ('timeout', -1), ('6', 6)] %}

; There should be one PTR record for each instance of the service you wish to advertise.
_nmos-registration._tcp	PTR	reg-api-{{ inst }}._nmos-registration._tcp
_nmos-register._tcp	PTR	reg-api-{{ inst }}._nmos-register._tcp
_nmos-query._tcp	PTR	qry-api-{{ inst }}._nmos-query._tcp

; Next we have a SRV and a TXT record corresponding to each PTR above, for the Registration API and Query API
; The SRV links the PTR name to a resolvable DNS name (see the A records above) and identify the port which the API runs on
; The TXT records indicate additional metadata relevant to the IS-04 spec

{% set port = reg_port_base + idx if inst != 'timeout' else 444 %}
{% set host = 'mocks' if inst != 'timeout' else 'timeout' %}
{% set ver = api_ver if inst != '1-ver' else 'v9.0' %}
{% set proto = api_proto if inst != '1-proto' else 'invalid' %}
{% set pri = loop.index0 * 10 %}

{% for api, srv in [('reg', 'registration'), ('reg', 'register'), ('qry', 'query')] %}

{{ api }}-api-{{ inst }}._nmos-{{ srv }}._tcp	SRV	0 0 {{ port }} {{ host }}.{{ domain }}.
{{ api }}-api-{{ inst }}._nmos-{{ srv }}._tcp	TXT	"api_ver={{ ver }}" "api_proto={{ proto }}" "pri={{ pri }}"

{% endfor %}

{% endfor %}
```

**Comment by andrewbonney** [comment]
I'd be a little concerned that if we need to make changes in the future we're more likely to make mistakes. If we're adding loops it may make more sense to loop over the three service types as they all use the same layout.

**Pull request merged**